File: Sources/YToolsCore/PanelCommandRouter.swift
public enum PanelInputMode: Sendable {
    case launcher
    case clipboard
}

public struct PanelKeyModifiers: OptionSet, Sendable {
    public let rawValue: UInt8

    public init(rawValue: UInt8) {
        self.rawValue = rawValue
    }

    public static let command = Self(rawValue: 1 << 0)
    public static let option = Self(rawValue: 1 << 1)
    public static let control = Self(rawValue: 1 << 2)
    public static let shift = Self(rawValue: 1 << 3)
}

public struct PanelKeyEvent: Sendable {
    public let keyCode: UInt16
    public let modifiers: PanelKeyModifiers

    public init(keyCode: UInt16, modifiers: PanelKeyModifiers) {
        self.keyCode = keyCode
        self.modifiers = modifiers
    }
}

public enum PanelCommand: Equatable, Sendable {
    case activateSelected
    case activateResult(Int)
    case addSelectedToBuffer(moveToNext: Bool)
    case removeLastBufferedItem
    case showFileBufferActions
    case clearFileBuffer
    case escape
    case moveSelection(Int)
    case showActions
    case navigateBack
    case deleteClipboardItem
    case saveClipboardAsSnippet
    case togglePreview
    case showLargeType
    case revealSelected
    case openSettings
}

public struct PanelCommandRouter: Sendable {
    private enum KeyCode {
        static let returnKey: UInt16 = 36
        static let leftArrow: UInt16 = 123
        static let rightArrow: UInt16 = 124
        static let downArrow: UInt16 = 125
        static let upArrow: UInt16 = 126
        static let delete: UInt16 = 51
        static let escape: UInt16 = 53
        static let d: UInt16 = 2
        static let s: UInt16 = 1
        static let y: UInt16 = 16
        static let l: UInt16 = 37
        static let comma: UInt16 = 43
        static let numberRow: [UInt16] = [18, 19, 20, 21, 23, 22, 26, 28, 25]
    }

    public init() {}

    public func command(for event: PanelKeyEvent, mode: PanelInputMode) -> PanelCommand? {
        if event.modifiers.contains(.command), mode == .launcher,
           let index = resultIndex(forNumberKeyCode: event.keyCode) {
            return .activateResult(index)
        }

        switch (event.keyCode, mode) {
        case (KeyCode.returnKey, _) where event.modifiers.isEmpty:
            return .activateSelected
        case (KeyCode.upArrow, .launcher) where event.modifiers.contains(.option):
            return .addSelectedToBuffer(moveToNext: false)
        case (KeyCode.downArrow, .launcher) where event.modifiers.contains(.option):
            return .addSelectedToBuffer(moveToNext: true)
        case (KeyCode.leftArrow, .launcher) where event.modifiers.contains(.option):
            return .removeLastBufferedItem
        case (KeyCode.rightArrow, .launcher) where event.modifiers.contains(.option):
            return .showFileBufferActions
        case (KeyCode.delete, .launcher) where event.modifiers.contains(.option):
            return .clearFileBuffer
        case (KeyCode.escape, _):
            return .escape
        case (KeyCode.downArrow, _):
            return .moveSelection(1)
        case (KeyCode.upArrow, _):
            return .moveSelection(-1)
        case (KeyCode.rightArrow, .launcher):
            return .showActions
        case (KeyCode.leftArrow, .launcher):
            return .navigateBack
        case (KeyCode.d, .clipboard) where event.modifiers.contains(.command):
            return .deleteClipboardItem
        case (KeyCode.s, .clipboard) where event.modifiers.contains(.command):
            return .saveClipboardAsSnippet
        case (KeyCode.y, .launcher) where event.modifiers.contains(.command):
            return .togglePreview
        case (KeyCode.l, .launcher) where event.modifiers.contains(.command):
            return .showLargeType
        case (KeyCode.returnKey, .launcher) where event.modifiers.contains(.command):
            return .revealSelected
        case (KeyCode.comma, _) where event.modifiers.contains(.command):
            return .openSettings
        default:
            return nil
        }
    }

    private func resultIndex(forNumberKeyCode keyCode: UInt16) -> Int? {
        KeyCode.numberRow.firstIndex(of: keyCode)
    }
}

File: src/YTools.Windows/Core/PanelCommandRouter.cs
namespace YTools.Core;

public enum PanelInputMode
{
    Launcher,
    Clipboard
}

[Flags]
public enum PanelKeyModifiers : byte
{
    None = 0,
    Command = 1 << 0,
    Option = 1 << 1,
    Control = 1 << 2,
    Shift = 1 << 3
}

public readonly record struct PanelKeyEvent(int KeyCode, PanelKeyModifiers Modifiers);

public enum PanelCommandKind
{
    ActivateSelected,
    ActivateResult,
    AddSelectedToBuffer,
    RemoveLastBufferedItem,
    ShowFileBufferActions,
    ClearFileBuffer,
    Escape,
    MoveSelection,
    ShowActions,
    NavigateBack,
    DeleteClipboardItem,
    SaveClipboardAsSnippet,
    TogglePreview,
    ShowLargeType,
    RevealSelected,
    OpenSettings
}

public readonly record struct PanelCommand(PanelCommandKind Kind, int Payload = 0);

/// <summary>
/// Key-table port using Win32 virtual-key codes, so the router stays testable
/// without a live window. Command maps to Ctrl, Option maps to Alt.
/// </summary>
public sealed class PanelCommandRouter
{
    private const int VkReturn = 0x0D;
    private const int VkLeft = 0x25;
    private const int VkRight = 0x27;
    private const int VkUp = 0x26;
    private const int VkDown = 0x28;
    private const int VkBack = 0x08;
    private const int VkEscape = 0x1B;
    private const int VkD = 0x44;
    private const int VkS = 0x53;
    private const int VkY = 0x59;
    private const int VkL = 0x4C;
    private const int VkOemComma = 0xBC;

    private static readonly int[] NumberRow =
    {
        0x30, 0x31, 0x32, 0x33, 0x34, 0x35, 0x36, 0x37, 0x38, 0x39
    };

    public PanelCommand? Command(PanelKeyEvent keyEvent, PanelInputMode mode)
    {
        if (keyEvent.Modifiers.HasFlag(PanelKeyModifiers.Command)
            && mode == PanelInputMode.Launcher
            && ResultIndexForNumberKeyCode(keyEvent.KeyCode) is { } index)
        {
            return new PanelCommand(PanelCommandKind.ActivateResult, index);
        }

        return (keyEvent.KeyCode, mode) switch
        {
            (VkReturn, _) when keyEvent.Modifiers == PanelKeyModifiers.None =>
                new PanelCommand(PanelCommandKind.ActivateSelected),
            (VkUp, PanelInputMode.Launcher) when keyEvent.Modifiers.HasFlag(PanelKeyModifiers.Option) =>
                new PanelCommand(PanelCommandKind.AddSelectedToBuffer, 0),
            (VkDown, PanelInputMode.Launcher) when keyEvent.Modifiers.HasFlag(PanelKeyModifiers.Option) =>
                new PanelCommand(PanelCommandKind.AddSelectedToBuffer, 1),
            (VkLeft, PanelInputMode.Launcher) when keyEvent.Modifiers.HasFlag(PanelKeyModifiers.Option) =>
                new PanelCommand(PanelCommandKind.RemoveLastBufferedItem),
            (VkRight, PanelInputMode.Launcher) when keyEvent.Modifiers.HasFlag(PanelKeyModifiers.Option) =>
                new PanelCommand(PanelCommandKind.ShowFileBufferActions),
            (VkBack, PanelInputMode.Launcher) when keyEvent.Modifiers.HasFlag(PanelKeyModifiers.Option) =>
                new PanelCommand(PanelCommandKind.ClearFileBuffer),
            (VkEscape, _) => new PanelCommand(PanelCommandKind.Escape),
            (VkDown, _) => new PanelCommand(PanelCommandKind.MoveSelection, 1),
            (VkUp, _) => new PanelCommand(PanelCommandKind.MoveSelection, -1),
            (VkRight, PanelInputMode.Launcher) => new PanelCommand(PanelCommandKind.ShowActions),
            (VkLeft, PanelInputMode.Launcher) => new PanelCommand(PanelCommandKind.NavigateBack),
            (VkD, PanelInputMode.Clipboard) when keyEvent.Modifiers.HasFlag(PanelKeyModifiers.Command) =>
                new PanelCommand(PanelCommandKind.DeleteClipboardItem),
            (VkS, PanelInputMode.Clipboard) when keyEvent.Modifiers.HasFlag(PanelKeyModifiers.Command) =>
                new PanelCommand(PanelCommandKind.SaveClipboardAsSnippet),
            (VkY, PanelInputMode.Launcher) when keyEvent.Modifiers.HasFlag(PanelKeyModifiers.Command) =>
                new PanelCommand(PanelCommandKind.TogglePreview),
            (VkL, PanelInputMode.Launcher) when keyEvent.Modifiers.HasFlag(PanelKeyModifiers.Command) =>
                new PanelCommand(PanelCommandKind.ShowLargeType),
            (VkReturn, PanelInputMode.Launcher) when keyEvent.Modifiers.HasFlag(PanelKeyModifiers.Command) =>
                new PanelCommand(PanelCommandKind.RevealSelected),
            (VkOemComma, _) when keyEvent.Modifiers.HasFlag(PanelKeyModifiers.Command) =>
                new PanelCommand(PanelCommandKind.OpenSettings),
            _ => null
        };
    }

    private static int? ResultIndexForNumberKeyCode(int keyCode)
    {
        var index = Array.IndexOf(NumberRow, keyCode);
        return index >= 0 ? index : null;
    }
}
